### PR TITLE
Uncomment --independent-goals flag.

### PR DESCRIPTION
**Pull request by grayjay**
--independent-goals could be useful for using local packages as setup or
build-tool dependencies in new-build.  See #4295 for an example of a dependency
problem that requires --independent-goals.  The flag is off by default.

I also had to update the constraints that cabal uses to force all build targets to be built with local packages.  They were too restrictive before, because they applied to all top-level dependencies.  Now they only apply to specific build targets.

@dcoutts Is there any problem with exposing this flag in new-build?  I didn't really understand the comments explaining why some of the flags were commented out in new-build, so I didn't know whether they still applied.  I marked them with Github comments.

### CODE REVIEW

File: cabal-install/Distribution/Client/ProjectPlanning.hs [hunk 1]
```diff
@@ -898,8 +898,7 @@ planPackages verbosity comp platform solver SolverSettings{..}
 
         setMaxBackjumps solverSettingMaxBackjumps
 
-        --TODO: [required eventually] should only be configurable for custom installs
-   -- . setIndependentGoals solverSettingIndependentGoals
+      . setIndependentGoals solverSettingIndependentGoals
```
Comment (grayjay) [inline cabal-install/Distribution/Client/ProjectPlanning.hs]: here

File: cabal-install/Distribution/Client/ProjectConfig/Types.hs [hunk 2]
```diff
@@ -360,10 +360,10 @@ data SolverSettings
        solverSettingCountConflicts    :: CountConflicts,
        solverSettingStrongFlags       :: StrongFlags,
        solverSettingAllowBootLibInstalls :: AllowBootLibInstalls,
-       solverSettingIndexState        :: Maybe IndexState
+       solverSettingIndexState        :: Maybe IndexState,
+       solverSettingIndependentGoals  :: IndependentGoals
        -- Things that only make sense for manual mode, not --local mode
        -- too much control!
-     --solverSettingIndependentGoals  :: IndependentGoals,
```
Comment (grayjay) [inline cabal-install/Distribution/Client/ProjectConfig/Types.hs]: here

File: cabal-install/Distribution/Client/ProjectConfig/Types.hs [hunk 3]
```diff
@@ -185,11 +185,11 @@ data ProjectConfigShared
        projectConfigCountConflicts    :: Flag CountConflicts,
        projectConfigStrongFlags       :: Flag StrongFlags,
        projectConfigAllowBootLibInstalls :: Flag AllowBootLibInstalls,
-       projectConfigPerComponent      :: Flag Bool
+       projectConfigPerComponent      :: Flag Bool,
+       projectConfigIndependentGoals  :: Flag IndependentGoals
 
        -- More things that only make sense for manual mode, not --local mode
        -- too much control!
-     --projectConfigIndependentGoals  :: Flag IndependentGoals,
```
Comment (grayjay) [inline cabal-install/Distribution/Client/ProjectConfig/Types.hs]: here

**Comment by ezyang** [comment]
Given the thought you've put into this, and the tests, I'm pretty OK with merging this.

**Comment by 23Skidoo** [comment]
Merged, thanks!

**Comment by grayjay** [comment]
Thanks!

**Pull request merged**